### PR TITLE
Apollo Client 2.0

### LINKED ISSUE #54: Apollo Client 2.0
Apollo Client 2.0 is just around the corner. Here's a general summary of how to migrate: https://github.com/apollographql/apollo-client/blob/master/Upgrade.md

I haven't yet studied it in detail, but there is likely a bit of work to do to keep ember-apollo-client up-to-date with it.

### PR DESCRIPTION
**Pull request by bgentry**
This PR attempts to upgrade the addon to support Apollo Client 2.0. Here is some background on what's required to upgrade an app which was previously using Apollo Client v1.0: https://www.apollographql.com/docs/react/2.0-migration.html

I've gotten the addon tests passing, and also got the tests for my app passing.

I have not yet updated any docs, nor have I finalized the config API for the Apollo service. Notably, the concept of "middlewares" has changed substantially with Apollo Client 2.0 with the addition of `apollo-link`. As of now, I've simply removed the `middlewares` config from the Apollo service because I'm not sure there's a way to cleanly migrate folks over to links.

I also took this opportunity to remove the deprecated `UnsubscribeRoute` mixin.

Fixes #54. This should also help with #65.

@dfreeman, if you have a chance to look over the webpack bundling changes I made, that would be super helpful :pray:

/cc @villander to test this out. 

<details>
<summary>
For those looking to test this out, here's my own app's updated Apollo service.
</summary>

```js
import { computed } from "@ember/object";
import { inject as service } from "@ember/service";
import ApolloService from "ember-apollo-client/services/apollo";
import { setContext } from "apollo-link-context";
import {
  InMemoryCache,
  IntrospectionFragmentMatcher,
} from "apollo-cache-inmemory";
import TypeIntrospectionQuery from "act/utils/graphql-type-query";
import { Promise as RSVPPromise } from "rsvp";

export default ApolloService.extend({
  cache: computed(function() {
    return new InMemoryCache({
      fragmentMatcher: new IntrospectionFragmentMatcher({
        introspectionQueryResultData: TypeIntrospectionQuery,
      }),
    });
  }),

  link: computed(function() {
    let httpLink = this._super(...arguments);

    let authLink = setContext((request, context) => {
      return this._runAuthorize(request, context);
    });
    return authLink.concat(httpLink);
  }),

  session: service(),

  _runAuthorize() {
    if (!this.get("session.isAuthenticated")) {
      return {};
    }
    return new RSVPPromise(success => {
      this.get(
        "session"
      ).authorize("authorizer:oauth2", (headerName, headerContent) => {
        let headers = {};
        headers[headerName] = headerContent;
        success({ headers });
      });
    });
  },
});
```
</details>

### CODE REVIEW

**Comment by villander** [comment]
@bgentry this works for me, but with a limitation. I can not add two middlewares / links

I followed this documentation, which passes an array of links, but does not work

https://www.apollographql.com/docs/react/basics/network-layer.html#linkMiddleware

```
link: computed(function() {
    let httpLink = this._super(...arguments);

    let authMiddleware = setContext((request, context) => {
      return this._runAuthorize(request, context);
    });

    let localeMiddleware = setContext((request, context) => {
      return this._setLocale(request, context);
    });

    return authMiddleware.concat(httpLink);
  }),
```

I nedd add `localeMiddleware`, when i add using `from()` the authentication stop works

**Comment by bgentry** [comment]
@villander can you post an example of the code that doesn't work? Actually if you can, a PR to this one that adds a failing test would be ideal. Maybe add a 2nd apollo service (`ApolloLinksService`) that uses multiple links and an acceptance test that exercises it via a route.

**Comment by dfreeman** [comment]
@bgentry Sorry, I've been away from a computer for the past day or so, but I should have a chance to look this over tomorrow :)

**Comment by bgentry** [comment]
Hi @dfreeman, just wanted to see if you'd gotten a chance to look at the bundling changes here. I'm keen to get this merged soon :)

**Comment by dfreeman** [comment]
@bgentry I unfortunately haven't had the opportunity to read up on the overall shape of Apollo Client 2.0 and the new packages it brings into play, but the changes you have seem reasonable to me within the scope of what I do understand 🙂

**Comment by josemarluedke** [comment]
I'm looking forward to use v2 of Apollo with Ember.

What are the blockers to get this merged?

**Comment by bgentry** [comment]
I’d like to get #83 merged and make sure we have a clear story for using multiple Apollo links (to replace the middleware concept). Also need to update a handful of documentation items.

Aside from that, I’ve been using this branch on my main production app for weeks with no issues. It certainly wouldn’t hurt to have more people test it out to cover some more use cases.

**Comment by xtagon** [comment]
👍 Also looking forward to it. I intend to test it out myself in the next couple weeks (time permitting).

**Comment by rachid** [comment]
Hi @bgentry nice going! I picked up the branch myself and currently having a working setup using an absinthe (phoenix) backend. Using absinthe-socket (apollo) running through sockets. I still have some unanswered issues concerning subscriptions but that needs another thread / chat.

**Comment by xtagon** [comment]
@rachid That's exciting, and I would love to be a part of the other thread/chat when you start it since my use case is exactly the same (Absinthe + subscriptions). 👍

**Comment by villander** [comment]
@bgentry now we need add support to multiple Apollo links right? How can we work on this together?

**Comment by bgentry** [comment]
@villander can you create a sample app that tries to use multiple links? If we have that plus an error that’s a good starting point for getting help from the Apollo team.

**Comment by villander** [comment]
Yes @bgentry i will do this tomorrow. And ping you when i finish

**Comment by ymartin** [comment]
Any updates on 2.0?

**Comment by bgentry** [comment]
This branch is ready to go afaik. The only issue is that the readme isn't yet updated, so it's probably going to be tough for folks to figure out how to use it.

I'm planning to go ahead and merge this in the next couple days now that Apollo Client 2.0 is required to use the Apollo DevTools (chrome inspector). Feel free to work off of this branch in the mean time.

File: addon/mixins/unsubscribe-route.js [hunk 1]
```diff
@@ -1,20 +0,0 @@
...
-    deprecate(`The \`UnsubscribeRoute\` mixin is deprecated, use \`RouteQueryManager\` instead.`, false, {
-      id: 'ember-apollo-client.deprecate-unsubscribe-route',
-      until: '1.0.0',
-    });
-
-    this._super(...arguments);
-
-    // If the model came from an apollo query, it will have an
-    // _apolloUnsubscribe function to stop it from receiving store updates.
-    let unsubscribe = this.get('controller.model._apolloUnsubscribe');
```
Comment (villander) [inline addon/mixins/unsubscribe-route.js]: @bgentry we must remove `subscription.unsubscribe()` from mixins as `ObjectQueryManager` right?

```js
_apolloUnsubscribe() {
    subscription.unsubscribe();
 },
```

**Comment by villander** [comment]
@bgentry this is done!

**Comment by villander** [comment]
This is a new ember app + fastboot working with this branch https://github.com/villander/ember-mvp

🚀

**Comment by bgentry** [comment]
One last fix I want to get in: #92.

Merging this PR as soon as I get that one in and the tests go green.

**Comment by bgentry** [comment]
v1.0.0-beta.1 released and published on npm! Thanks all for the help, especially @villander.

I'm going to leave this branch around for a while to avoid breaking others who may be pointing at it.

**Comment by xtagon** [comment]
Big thanks to all of you for doing this!! 👍 

@rachid Did you figure out anything regarding the Absinthe subscriptions questions you had? If it was moved to another discussion I would be curious to see what came of it since that's the first thing I'm going to be trying

**Pull request merged**